drivers/at_cmd/at_cmd.c

```diff
@@ -26,13 +27,22 @@ static K_THREAD_STACK_DEFINE(socket_thread_stack, \
 				CONFIG_AT_CMD_THREAD_STACK_SIZE);
 
 static K_SEM_DEFINE(cmd_pending, 1, 1);
+static K_SEM_DEFINE(list_pending, 1, 1);
+
+/**@brief Link list element for notification handler. */
+typedef struct {
+	sys_snode_t         node;
+	at_cmd_handler_t    handler;
+} notif_handler_t;
 
 static int              common_socket_fd;
 static char             *response_buf;
 static u32_t            response_buf_len;
 
 static struct k_thread  socket_thread;
-static at_cmd_handler_t notification_handler;
+static sys_slist_t      handler_list;
```

Comment:
I'd call it `notification_handler_list` to explicitly differentiate it from command handler.